op-program/host/l2_source.go

```
@@ -0,0 +1,139 @@
+	if l.experimentalEnabled {
+		// This means experimental source was not able to retrieve relevant information from eth_getProof or debug_executionWitness
+		// We should fall back to the canonical source, and log a warning, and record a metric
+		l.logger.Warn("Experimental source failed to retrieve code by hash, falling back to canonical source", "hash", hash)
+	}
+	return l.canonicalDebugClient.CodeByHash(ctx, hash)
+}
+
+// NodeByHash implements prefetcher.L2Source.
+func (l *L2Source) NodeByHash(ctx context.Context, hash common.Hash) ([]byte, error) {
+	if l.experimentalEnabled {
+		// This means experimental source was not able to retrieve relevant information from eth_getProof or debug_executionWitness
+		// We should fall back to the canonical source, and log a warning, and record a metric
+		l.logger.Warn("Experimental source failed to retrieve node by hash, falling back to canonical source", "hash", hash)
+	}
+	return l.canonicalDebugClient.NodeByHash(ctx, hash)
+}
+
+// InfoAndTxsByHash implements prefetcher.L2Source.
+func (l *L2Source) InfoAndTxsByHash(ctx context.Context, blockHash common.Hash) (eth.BlockInfo, types.Transactions, error) {
+	if l.experimentalEnabled {
+		return l.experimentalClient.InfoAndTxsByHash(ctx, blockHash)
+	}
+	return l.canonicalEthClient.InfoAndTxsByHash(ctx, blockHash)
+}
+
+// OutputByRoot implements prefetcher.L2Source.
+func (l *L2Source) OutputByRoot(ctx context.Context, root common.Hash) (eth.Output, error) {
+	if l.experimentalEnabled {
+		return l.experimentalClient.OutputByRoot(ctx, root)
+	}
+	return l.canonicalEthClient.OutputByRoot(ctx, root)
+}
+
+// ExecutionWitness implements prefetcher.L2Source.
+func (l *L2Source) ExecutionWitness(ctx context.Context, blockNum uint64) (*eth.ExecutionWitness, error) {
+	if !l.experimentalEnabled {
+		l.logger.Error("Experimental source is not enabled, cannot fetch execution witness", "blockNum", blockNum)
+		panic("experimental source is not enabled")
```

I suspect this should just return an `ErrNotEnabled` rather than logging and panicing. Looking at https://github.com/ethereum-optimism/optimism/pull/12559/ we'll always get the new hint whether this is enabled or not, and the fetcher will have to handle not having witness support enabled.

I know that's a little different to my previous comment about the caller not calling this at all, but I think it probably works out cleaner to just check for a specific error rather than having to pass a `witnessEnabled` bool through to the prefetcher. The prefetcher would just return a `nil` error if it got `ErrNotEnabled` back without storing anything and the fallback to debug_dbGet would automatically kick in as needed.

---

fixed here: fee2752 (handling for ignoring this error in the prefetcher added in next PR because that's where the logic was added)